### PR TITLE
Add Country custom select field to Optin campaign

### PR DESCRIPTION
**Pull request by ShoboySnr**
Adding country custom select field to the optin campaign

### CODE REVIEW

File: src/OptinForms/AbstractOptinTheme.php [hunk 1]
```diff
@@ -1121,6 +1121,44 @@ public function shortcode_optin_form_custom_fields($atts)
...
+                                    }
+                                break;
+                                case 'alpha_three':
+                                    $countries_lists =  \MailOptin\Core\countries_array('alpha_three');
+                                    if(is_array($countries_lists) && !empty($countries_lists)) {
+                                        $placeholder = ! empty($placeholder) ? $placeholder : esc_html__('Select...', 'mailoptin');
+                                        $html        .= "<select name=\"$name_attribute\" $data_attr class=\"$class\" id=\"$id\" style=\"$style\">";
+                                        $html        .= "<option value='' selected='selected'>$placeholder</option>";
+
+                                        foreach ($countries_lists as $index => $countries_list) {
```
Comment (w3guy) [inline src/OptinForms/AbstractOptinTheme.php:1153]: `$index` is already used by the outer foreach loop. Change to maybe `$key`.

File: src/OptinForms/AbstractOptinTheme.php [hunk 2]
```diff
@@ -1121,6 +1121,44 @@ public function shortcode_optin_form_custom_fields($atts)
...
+                                            $html .= sprintf('<option value="%s (%s)">%s</option>', $countries_list, $index, $index);
```
Comment (w3guy) [inline src/OptinForms/AbstractOptinTheme.php:1154]: See? Now you have two $index variable. So which belongs to the outer and which is the inner one?

File: src/OptinForms/AbstractOptinTheme.php [hunk 3]
```diff
@@ -1121,6 +1121,44 @@ public function shortcode_optin_form_custom_fields($atts)
...
+                                        }
+                                        $html .= '</select>';
+                                    }
+                                break;
+                                case 'alpha_three':
+                                    $countries_lists =  \MailOptin\Core\countries_array('alpha_three');
+                                    if(is_array($countries_lists) && !empty($countries_lists)) {
+                                        $placeholder = ! empty($placeholder) ? $placeholder : esc_html__('Select...', 'mailoptin');
+                                        $html        .= "<select name=\"$name_attribute\" $data_attr class=\"$class\" id=\"$id\" style=\"$style\">";
```
Comment (w3guy) [inline src/OptinForms/AbstractOptinTheme.php:1150]: Notice: Undefined variable: name_attribute.

File: src/OptinForms/AbstractOptinTheme.php (hunk 2 above)
Comment (w3guy) [inline src/OptinForms/AbstractOptinTheme.php:1154]: And I think you got the sprinf parameters wrong. Here is the output

image

File: src/Admin/Customizer/CustomControls/WP_Customize_Fields_Repeater_Control.php [hunk 4]
```diff
@@ -589,13 +609,20 @@ public function template($index = 9999999999999)
...
+            'country'            => __('Country', 'mailoptin'),
         ];
 
         $widget_title = sprintf(__('Field %s', 'mailoptin'), '#' . ($index + 1));
         if (isset($this->saved_values[$index]['placeholder'])) {
             $widget_title = $this->saved_values[$index]['placeholder'];
         }
 
+        $list_country_name_select_type = [
+            ''  =>   __('Select One..', 'mailoptin'),
```
Comment (w3guy) [inline src/Admin/Customizer/CustomControls/WP_Customize_Fields_Repeater_Control.php:621]: Remove this. Let it always default to alpha-2.

File: src/Functions/countries.php [hunk 5]
```diff
@@ -9,254 +9,513 @@
...
+            'NIU' => __('Niue', 'mailoptin'),
+            'NFK' => __('Norfolk Island', 'mailoptin'),
+            'MNP' => __('Northern Mariana Islands', 'mailoptin'),
+            'PRK' => __('North Korea', 'mailoptin'),
+            'NOR' => __('Norway', 'mailoptin'),
+            'OMN' => __('Oman', 'mailoptin'),
+            'PAK' => __('Pakistan', 'mailoptin'),
+            'PSE' => __('Palestinian Territory', 'mailoptin'),
+            'PAN' => __('Panama', 'mailoptin'),
+            'PAN' => __('Papua New Guinea', 'mailoptin'),
```
Comment (w3guy) [inline src/Functions/countries.php:185]: Duplicate array key `PAN`.

Please always go over your work to ensure accuracy.

File: src/Admin/Customizer/CustomControls/WP_Customize_Fields_Repeater_Control.php [hunk 6]
```diff
@@ -638,6 +665,7 @@ public function template($index = 9999999999999)
                     <?php $this->repeater_textarea_field($index, 'field_options', '', __('Options', 'mailoptin'), __('Enter a comma-separated list of options', 'mailoptin')); ?>
                     <?php $this->repeater_text_field($index, 'hidden_value', '', __('Value', 'mailoptin'), __('Enter the value for this hidden field', 'mailoptin')); ?>
 
+                    <?php $this->repeater_select_field($index, 'list_country_options_field_options', $list_country_name_select_type, '', __('Show Country In', 'mailoptin')); ?>
```
Comment (w3guy) [inline src/Admin/Customizer/CustomControls/WP_Customize_Fields_Repeater_Control.php:668]: Change every instances of `list_country_options_field_options` to `country_field_options`

File: src/OptinForms/AbstractOptinTheme.php (hunk 2 above)
Comment (ShoboySnr) [reply src/OptinForms/AbstractOptinTheme.php:1154]: These two index are the same though as the second one will set the value of the third `%s`, but I'll still change it to `$key`

Comment (ShoboySnr) [reply src/OptinForms/AbstractOptinTheme.php:1154]: What should be the output? I was thinking of putting the full country as value and then the selected option inside a bracket. What do you think?

File: src/Admin/Customizer/CustomControls/WP_Customize_Fields_Repeater_Control.php (hunk 4 above)
Comment (ShoboySnr) [reply src/Admin/Customizer/CustomControls/WP_Customize_Fields_Repeater_Control.php:621]: Okay

**Comment by w3guy** [comment]
Maybe you should use named sprintf parameters eg $2%s to avoid duplicate
$index variable. And please change it to $key. It was flagged by my IDE.

**Comment by ShoboySnr** [comment]
Okay, got it

File: src/OptinForms/AbstractOptinTheme.php (hunk 2 above)
Comment (w3guy) [reply src/OptinForms/AbstractOptinTheme.php:1154]: The alpha codes should be the country value. while the country name and alpha code in bracket the label
The inverse is currently the case.

File: src/OptinForms/AbstractOptinTheme.php [hunk 7]
```diff
@@ -1121,6 +1121,41 @@ public function shortcode_optin_form_custom_fields($atts)
...
+                            switch ($display_type) {
+                                case 'alpha-2':
+                                    $countries_lists = \MailOptin\Core\countries_array();
+                                    if (is_array($countries_lists) && ! empty($countries_lists)) {
+                                        $placeholder = ! empty($placeholder) ? $placeholder : esc_html__('Select...', 'mailoptin');
+                                        $html        .= "<select name=\"$field_id\" $data_attr class=\"$class\" id=\"$id\" style=\"$style\">";
+                                        $html        .= "<option value='' selected='selected'>$placeholder</option>";
+
+                                        foreach ($countries_lists as $key => $countries_list) {
+                                            $html .= sprintf('<option value="%1$s">%2$s (%1$s)</option>', $key, $countries_list);
```
Comment (w3guy) [inline src/OptinForms/AbstractOptinTheme.php:1138]: @ShoboySnr this is how sprintf named/index parameter works. Notice there is no ending $key variable again?

Comment (ShoboySnr) [reply src/OptinForms/AbstractOptinTheme.php:1138]: Great.. thank you, understood

**Pull request merged**